File: python/localmodel.py
"""A local language model the app runs itself, for filing summaries — automatic and
transparent: the user installs nothing and types no commands.

Resolution order, cheapest first:
1. An endpoint the user already runs — `BAGHOLDER_LLM_URL`, or Ollama on its default
   port. If one answers, it is used and nothing is downloaded.
2. Otherwise the app provisions its own: it downloads a single self-contained model
   executable (a llamafile: model plus runtime in one portable file, Mac/Linux/
   Windows) into the app's home, verifies it against a pinned SHA-256, and runs it
   as a background server. This happens once, lazily, the first time a summary is
   asked for; later starts reuse the downloaded file.

Everything is best-effort and returns "" rather than raising, so a slow download, a
blocked binary, or no network simply means summaries are quiet — never a crash.

Security: the executable is fetched over HTTPS from a pinned Hugging Face URL and is
refused unless its SHA-256 matches the pin, so a tampered or truncated download is
never run. Override the artifact with `BAGHOLDER_LLAMAFILE_URL` /
`BAGHOLDER_LLAMAFILE_SHA256`, or point at your own server with `BAGHOLDER_LLM_URL`.
"""
from __future__ import annotations

import atexit
import hashlib
import json
import os
import shutil
import ssl
import stat
import subprocess
import sys
import threading
import time
from pathlib import Path
from urllib.request import Request, urlopen

# A small instruction-tuned model, enough for a one-sentence summary. Pinned by
# SHA-256 so only this exact file is ever executed.
LLAMAFILE_URL = os.environ.get(
    "BAGHOLDER_LLAMAFILE_URL",
    "https://huggingface.co/Mozilla/Llama-3.2-1B-Instruct-llamafile/resolve/main/Llama-3.2-1B-Instruct-Q4_K_M.llamafile",
)
LLAMAFILE_SHA256 = os.environ.get(
    "BAGHOLDER_LLAMAFILE_SHA256",
    "ac1c2864000bad7f62ee56ee908d3f55dd051a267d015b15fa6e831e69767b55",
)
USER_LLM_URL = os.environ.get("BAGHOLDER_LLM_URL", "").rstrip("/")
OLLAMA_URL = os.environ.get("BAGHOLDER_OLLAMA_URL", "http://127.0.0.1:11434").rstrip("/")
OLLAMA_MODEL = os.environ.get("BAGHOLDER_OLLAMA_MODEL", "llama3.2")
MANAGED_HOST = "127.0.0.1"
MANAGED_PORT = int(os.environ.get("BAGHOLDER_LLM_PORT", "8121"))
DOWNLOAD_TIMEOUT = 60 * 30       # a big file over a slow link
START_TIMEOUT = 120             # the server loading the model
CHAT_TIMEOUT = float(os.environ.get("BAGHOLDER_LLM_CHAT_TIMEOUT", "40"))
_ALLOWED_HOSTS = ("huggingface.co", "cdn-lfs.huggingface.co", "cdn-lfs-us-1.huggingface.co")

_lock = threading.Lock()
_state = {"phase": "off", "detail": "", "proc": None, "endpoint": "", "model": ""}
# phase: off | detecting | downloading | starting | ready | failed


def _home():
    return Path(os.environ.get("BAGHOLDER_HOME") or (Path.home() / ".bagholder"))


def _models_dir():
    import store                         # a test never makes or fills the person's model folder
    store.guard_home(_home())
    d = _home() / "models"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _llamafile_path():
    return _models_dir() / "summarizer.llamafile"


def _ctx():
    try:
        import certifi
        return ssl.create_default_context(cafile=certifi.where())
    except Exception:
        for ca in ("/etc/ssl/cert.pem", "/etc/ssl/certs/ca-certificates.crt"):
            if os.path.exists(ca):
                return ssl.create_default_context(cafile=ca)
    return ssl.create_default_context()


def _get_ok(url, timeout=3):
    try:
        urlopen(Request(url, headers={"User-Agent": "Bagholder"}), timeout=timeout, context=_ctx()).read()
        return True
    except Exception:
        return False


# --------------------------------------------------------------------------- #
# Endpoint resolution
# --------------------------------------------------------------------------- #
def _detect_running():
    """A user-run endpoint, if one answers now. Returns (base_url, model) or None."""
    if USER_LLM_URL and _get_ok(USER_LLM_URL + "/v1/models", 2):
        return USER_LLM_URL, os.environ.get("BAGHOLDER_LLM_MODEL", "local")
    if _get_ok(OLLAMA_URL + "/api/tags", 2):
        return OLLAMA_URL, OLLAMA_MODEL
    return None


def status():
    """One of: off, detecting, downloading, starting, ready, failed — for the UI."""
    with _lock:
        if _state["endpoint"]:
            return "ready"
        return _state["phase"]


def available():
    return bool(endpoint())


COMING_UP = ("detecting", "starting")     # phases that finish in seconds, unlike a download


def wait_ready(seconds):
    """Wait, for at most `seconds`, for a model that is coming up right now, and say whether
    one is up. Asking for the first time is what starts it, so without this the first caller
    of a session always gets nothing. A download is never waited for: that takes minutes and
    the caller has a page to answer."""
    endpoint()                            # the ask that starts one, if none is up
    deadline = time.time() + max(0.0, seconds)
    while time.time() < deadline:
        if available():
            return True
        if status() not in COMING_UP:
            return False
        time.sleep(0.5)
    return available()


def endpoint():
    """The base URL of a working local model, or "" if none is up yet. Never blocks
    on a download; if provisioning is needed, kick it in the background and return "".
    """
    with _lock:
        if _state["endpoint"]:
            return _state["endpoint"]
    found = _detect_running()
    if found:
        with _lock:
            _state["endpoint"], _state["model"], _state["phase"] = found[0], found[1], "ready"
        return found[0]
    ensure()
    return ""


# --------------------------------------------------------------------------- #
# Provisioning (background)
# --------------------------------------------------------------------------- #
def ensure():
    """Start provisioning the managed model if it is not already under way. Returns
    at once; progress is visible through status()."""
    with _lock:
        if _state["phase"] in ("detecting", "downloading", "starting") or _state["endpoint"]:
            return
        _state["phase"] = "detecting"
    threading.Thread(target=_provision, name="bagholder-localmodel", daemon=True).start()


def _set(phase, detail=""):
    with _lock:
        _state["phase"], _state["detail"] = phase, detail


def _provision():
    try:
        found = _detect_running()
        if found:
            with _lock:
                _state["endpoint"], _state["model"], _state["phase"] = found[0], found[1], "ready"
            return
        path = _llamafile_path()
        if not _verified(path):
            _set("downloading")
            if not _download(path):
                _set("failed", "download failed")
                return
        if not _verified(path):
            _set("failed", "checksum mismatch")
            try:
                path.unlink()
            except OSError:
                pass
            return
        _set("starting")
        if _spawn(path) and _wait_ready():
            with _lock:
                _state["endpoint"] = "http://%s:%d" % (MANAGED_HOST, MANAGED_PORT)
                _state["model"] = "local"
                _state["phase"] = "ready"
        else:
            _set("failed", "server did not start")
    except Exception as e:
        _set("failed", str(e))


def _verified(path):
    if not path.exists() or not LLAMAFILE_SHA256:
        return False
    h = hashlib.sha256()
    try:
        with open(path, "rb") as fh:
            for chunk in iter(lambda: fh.read(1 << 20), b""):
                h.update(chunk)
    except OSError:
        return False
    return h.hexdigest() == LLAMAFILE_SHA256.lower()


def _download(path):
    from urllib.parse import urlparse
    if urlparse(LLAMAFILE_URL).hostname not in _ALLOWED_HOSTS:
        return False
    tmp = path.with_suffix(".part")
    try:
        req = Request(LLAMAFILE_URL, headers={"User-Agent": "Bagholder"})
        with urlopen(req, timeout=DOWNLOAD_TIMEOUT, context=_ctx()) as r, open(tmp, "wb") as out:
            shutil.copyfileobj(r, out, length=1 << 20)
        tmp.replace(path)
        path.chmod(path.stat().st_mode | stat.S_IXUSR | stat.S_IXGRP)
        # a downloaded executable is quarantined on macOS; clear it so it can run
        if sys.platform == "darwin":
            try:
                subprocess.run(["xattr", "-d", "com.apple.quarantine", str(path)], capture_output=True, timeout=10)
            except Exception:
                pass
        return True
    except Exception:
        try:
            tmp.unlink()
        except OSError:
            pass
        return False


def _spawn(path):
    try:
        proc = subprocess.Popen(
            [str(path), "--server", "--nobrowser", "--host", MANAGED_HOST, "--port", str(MANAGED_PORT),
             "-ngl", "0", "--log-disable"],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, stdin=subprocess.DEVNULL,
        )
    except Exception:
        # some hosts must run the APE via a shell
        try:
            proc = subprocess.Popen(
                ["sh", str(path), "--server", "--nobrowser", "--host", MANAGED_HOST, "--port", str(MANAGED_PORT), "--log-disable"],
                stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, stdin=subprocess.DEVNULL,
            )
        except Exception:
            return False
    with _lock:
        _state["proc"] = proc
    return True


def _wait_ready():
    base = "http://%s:%d" % (MANAGED_HOST, MANAGED_PORT)
    deadline = time.time() + START_TIMEOUT
    while time.time() < deadline:
        with _lock:
            proc = _state["proc"]
        if proc and proc.poll() is not None:
            return False
        if _get_ok(base + "/health", 2) or _get_ok(base + "/v1/models", 2):
            return True
        time.sleep(2)
    return False


def shutdown():
    with _lock:
        proc = _state["proc"]
        _state["proc"] = None
    if proc and proc.poll() is None:
        try:
            proc.terminate()
            proc.wait(timeout=5)
        except Exception:
            try:
                proc.kill()
            except Exception:
                pass


atexit.register(shutdown)


# --------------------------------------------------------------------------- #
# Chat
# --------------------------------------------------------------------------- #
def chat(prompt, max_tokens=90):
    """One completion from the local model (OpenAI-compatible /v1/chat/completions,
    which both Ollama and llamafile speak), or "" on any failure or if no model is up
    yet. Kicks provisioning when nothing is running."""
    base = endpoint()
    if not base:
        return ""
    with _lock:
        model = _state["model"] or "local"
    body = json.dumps({
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.1,
        "max_tokens": int(max_tokens),
        "stream": False,
    }).encode("utf-8")
    try:
        req = Request(base + "/v1/chat/completions", data=body, headers={"Content-Type": "application/json"})
        resp = json.loads(urlopen(req, timeout=CHAT_TIMEOUT, context=_ctx()).read().decode("utf-8", "replace"))
        return str(((resp.get("choices") or [{}])[0].get("message") or {}).get("content") or "").strip()
    except Exception:
        return ""

File: rust/crates/market/src/localmodel.rs
//! A local language model the app runs itself, for filing summaries.
//!
//! Resolution order, cheapest first:
//! 1. An endpoint the user already runs -- `BAGHOLDER_LLM_URL`, or Ollama on
//!    its default port. If one answers, it is used and nothing is downloaded.
//! 2. Otherwise the app provisions its own: it downloads a single
//!    self-contained model executable (a llamafile) into the app's home,
//!    verifies it against a pinned SHA-256, and runs it as a background
//!    server. This happens once, lazily, the first time a summary is asked
//!    for; later starts reuse the downloaded file.
//!
//! Everything is best-effort and answers "" rather than failing. The file is
//! fetched over HTTPS from a pinned Hugging Face URL and refused unless its
//! SHA-256 matches the pin, so a tampered or truncated download is never run.

use serde_json::{json, Value};
use std::io::Read;
use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

/// Stand-ins, per thread, for what reaches outside the process: the running
/// endpoint probe, provisioning, the chat request and the wait's sleep.
pub mod hooks {
    use std::cell::RefCell;
    pub type Detect = Box<dyn Fn() -> Option<(String, String)>>;
    pub type Ensure = Box<dyn Fn()>;
    pub type Endpoint = Box<dyn Fn() -> String>;
    pub type Chat = Box<dyn Fn(&str, i64) -> String>;
    /// (url, body) -> response text, or Err for a backend error
    pub type Post = Box<dyn Fn(&str, &str) -> Result<String, String>>;
    pub type Available = Box<dyn Fn() -> bool>;
    pub type Status = Box<dyn Fn() -> &'static str>;
    thread_local! {
        pub static DETECT: RefCell<Option<Detect>> = RefCell::new(None);
        pub static ENSURE: RefCell<Option<Ensure>> = RefCell::new(None);
        pub static ENDPOINT: RefCell<Option<Endpoint>> = RefCell::new(None);
        pub static CHAT: RefCell<Option<Chat>> = RefCell::new(None);
        pub static POST: RefCell<Option<Post>> = RefCell::new(None);
        pub static AVAILABLE: RefCell<Option<Available>> = RefCell::new(None);
        pub static STATUS: RefCell<Option<Status>> = RefCell::new(None);
        pub static NO_SLEEP: RefCell<bool> = RefCell::new(false);
    }
    pub fn clear() {
        DETECT.with(|h| *h.borrow_mut() = None);
        ENSURE.with(|h| *h.borrow_mut() = None);
        ENDPOINT.with(|h| *h.borrow_mut() = None);
        CHAT.with(|h| *h.borrow_mut() = None);
        POST.with(|h| *h.borrow_mut() = None);
        AVAILABLE.with(|h| *h.borrow_mut() = None);
        STATUS.with(|h| *h.borrow_mut() = None);
        NO_SLEEP.with(|h| *h.borrow_mut() = false);
    }
}

/// Back to the state a fresh process starts in (nothing detected, off).
pub fn reset_state() {
    let mut st = state().lock().unwrap();
    st.phase = "off";
    st.detail.clear();
    st.proc = None;
    st.endpoint.clear();
    st.model.clear();
}

fn env(name: &str, default: &str) -> String {
    std::env::var(name).unwrap_or_else(|_| default.to_string())
}

pub fn llamafile_url() -> String {
    env("BAGHOLDER_LLAMAFILE_URL", "https://huggingface.co/mozilla-ai/gemma-2-2b-it-llamafile/resolve/main/gemma-2-2b-it.Q4_K_M.llamafile")
}

pub fn llamafile_sha256() -> String {
    env("BAGHOLDER_LLAMAFILE_SHA256", "5eae1115b231c9115b260cc2442263db9e84dec99be8610cfe19fac137284217")
}

fn user_llm_url() -> String {
    env("BAGHOLDER_LLM_URL", "").trim_end_matches('/').to_string()
}

fn ollama_url() -> String {
    env("BAGHOLDER_OLLAMA_URL", "http://127.0.0.1:11434").trim_end_matches('/').to_string()
}

fn ollama_model() -> String {
    env("BAGHOLDER_OLLAMA_MODEL", "llama3.2")
}

pub const MANAGED_HOST: &str = "127.0.0.1";

fn managed_port() -> u16 {
    env("BAGHOLDER_LLM_PORT", "8121").parse().unwrap_or(8121)
}

/// A big file over a slow link.
pub const DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(60 * 30);
/// The server loading the model.
pub const START_TIMEOUT: Duration = Duration::from_secs(120);

fn chat_timeout() -> Duration {
    Duration::from_secs_f64(env("BAGHOLDER_LLM_CHAT_TIMEOUT", "40").parse().unwrap_or(40.0))
}

const ALLOWED_HOSTS: [&str; 3] = ["huggingface.co", "cdn-lfs.huggingface.co", "cdn-lfs-us-1.huggingface.co"];

/// Phases that finish in seconds, unlike a download.
pub const COMING_UP: [&str; 2] = ["detecting", "starting"];

struct State {
    phase: &'static str,
    detail: String,
    proc: Option<Child>,
    endpoint: String,
    model: String,
}

fn state() -> &'static Mutex<State> {
    static S: OnceLock<Mutex<State>> = OnceLock::new();
    S.get_or_init(|| Mutex::new(State { phase: "off", detail: String::new(), proc: None, endpoint: String::new(), model: String::new() }))
}

fn home() -> PathBuf {
    match std::env::var("BAGHOLDER_HOME") {
        Ok(h) if !h.is_empty() => PathBuf::from(h),
        _ => PathBuf::from(std::env::var("HOME").unwrap_or_default()).join(".bagholder-rust"),
    }
}

fn llamafile_path() -> PathBuf {
    // a test never makes or fills the person's model folder
    let d = bagholder_store::guard_home(&home()).expect("a test reached the real data folder").join("models");
    let _ = std::fs::create_dir_all(&d);
    d.join("summarizer.llamafile")
}

fn get_ok(url: &str, timeout: Duration) -> bool {
    bagholder_net::client::request("GET", url, &[("User-Agent", "Bagholder")], None, timeout).is_ok()
}

/// A user-run endpoint, if one answers now.
fn detect_running() -> Option<(String, String)> {
    if let Some(r) = hooks::DETECT.with(|h| h.borrow().as_ref().map(|f| f())) {
        return r;
    }
    let user = user_llm_url();
    if !user.is_empty() && get_ok(&format!("{}/v1/models", user), Duration::from_secs(2)) {
        return Some((user, env("BAGHOLDER_LLM_MODEL", "local")));
    }
    let ollama = ollama_url();
    if get_ok(&format!("{}/api/tags", ollama), Duration::from_secs(2)) {
        return Some((ollama, ollama_model()));
    }
    None
}

/// Off, detecting, downloading, starting, ready, failed.
pub fn status() -> &'static str {
    if let Some(r) = hooks::STATUS.with(|h| h.borrow().as_ref().map(|f| f())) {
        return r;
    }
    let st = state().lock().unwrap();
    if !st.endpoint.is_empty() { "ready" } else { st.phase }
}

pub fn available() -> bool {
    if let Some(r) = hooks::AVAILABLE.with(|h| h.borrow().as_ref().map(|f| f())) {
        return r;
    }
    !endpoint().is_empty()
}

/// Whether a model is up now, as already known: asks nothing and starts nothing,
/// so it can be consulted as often as anyone likes.
pub fn is_ready() -> bool {
    if let Some(r) = hooks::AVAILABLE.with(|h| h.borrow().as_ref().map(|f| f())) {
        return r;
    }
    !state().lock().unwrap().endpoint.is_empty()
}

/// Wait at most `seconds` for a model that is coming
/// up right now. A download is never waited for.
pub fn wait_ready(seconds: f64) -> bool {
    endpoint();
    let deadline = Instant::now() + Duration::from_secs_f64(seconds.max(0.0));
    while Instant::now() < deadline {
        if available() {
            return true;
        }
        if !COMING_UP.contains(&status()) {
            return false;
        }
        if !hooks::NO_SLEEP.with(|h| *h.borrow()) {
            std::thread::sleep(Duration::from_millis(500));
        }
    }
    available()
}

/// The base URL of a working local model, or "" if none
/// is up yet. Never blocks on a download.
pub fn endpoint() -> String {
    if let Some(r) = hooks::ENDPOINT.with(|h| h.borrow().as_ref().map(|f| f())) {
        return r;
    }
    {
        let st = state().lock().unwrap();
        if !st.endpoint.is_empty() {
            return st.endpoint.clone();
        }
    }
    if let Some((url, model)) = detect_running() {
        let mut st = state().lock().unwrap();
        st.endpoint = url.clone();
        st.model = model;
        st.phase = "ready";
        drop(st);
        changed();
        return url;
    }
    ensure();
    String::new()
}

/// Start provisioning if it is not already under way.
pub fn ensure() {
    if hooks::ENSURE.with(|h| h.borrow().as_ref().map(|f| f()).is_some()) {
        return;
    }
    {
        let mut st = state().lock().unwrap();
        if ["detecting", "downloading", "starting"].contains(&st.phase) || !st.endpoint.is_empty() {
            return;
        }
        st.phase = "detecting";
    }
    let _ = std::thread::Builder::new().name("bagholder-localmodel".into()).spawn(provision);
}

static ON_CHANGE: OnceLock<Box<dyn Fn() + Send + Sync>> = OnceLock::new();

/// Called whenever the model's phase changes (coming up, ready, failed): how whoever
/// waits for a model learns of it without asking again and again. Set once, at start.
pub fn on_change(f: impl Fn() + Send + Sync + 'static) {
    let _ = ON_CHANGE.set(Box::new(f));
}

fn changed() {
    if let Some(f) = ON_CHANGE.get() {
        f();
    }
}

fn set(phase: &'static str, detail: &str) {
    {
        let mut st = state().lock().unwrap();
        st.phase = phase;
        st.detail = detail.to_string();
    }
    changed();
}

fn provision() {
    if let Some((url, model)) = detect_running() {
        let mut st = state().lock().unwrap();
        st.endpoint = url;
        st.model = model;
        st.phase = "ready";
        drop(st);
        changed();
        return;
    }
    let path = llamafile_path();
    if !verified(&path) {
        set("downloading", "");
        if !download(&path) {
            set("failed", "download failed");
            return;
        }
    }
    if !verified(&path) {
        set("failed", "checksum mismatch");
        let _ = std::fs::remove_file(&path);
        return;
    }
    set("starting", "");
    if spawn(&path) && wait_started() {
        let mut st = state().lock().unwrap();
        st.endpoint = format!("http://{}:{}", MANAGED_HOST, managed_port());
        st.model = "local".into();
        st.phase = "ready";
        drop(st);
        changed();
    } else {
        set("failed", "server did not start");
    }
}

pub fn verified(path: &PathBuf) -> bool {
    let pin = llamafile_sha256();
    if !path.exists() || pin.is_empty() {
        return false;
    }
    let mut f = match std::fs::File::open(path) { Ok(f) => f, Err(_) => return false };
    let mut h = openssl::sha::Sha256::new();
    let mut buf = vec![0u8; 1 << 20];
    loop {
        match f.read(&mut buf) {
            Ok(0) => break,
            Ok(n) => h.update(&buf[..n]),
            Err(_) => return false,
        }
    }
    let hex: String = h.finish().iter().map(|b| format!("{:02x}", b)).collect();
    hex == pin.to_lowercase()
}

pub fn download(path: &PathBuf) -> bool {
    let url = llamafile_url();
    let host = url.split("://").nth(1).and_then(|r| r.split('/').next()).unwrap_or("").split(':').next().unwrap_or("").to_string();
    if !ALLOWED_HOSTS.contains(&host.as_str()) {
        return false;
    }
    let tmp = path.with_extension("part");
    let got = bagholder_net::client::request("GET", &url, &[("User-Agent", "Bagholder")], None, DOWNLOAD_TIMEOUT)
        .map_err(|e| e.to_string())
        .and_then(|r| std::fs::write(&tmp, &r.body).map_err(|e| e.to_string()))
        .and_then(|_| std::fs::rename(&tmp, path).map_err(|e| e.to_string()));
    if got.is_err() {
        let _ = std::fs::remove_file(&tmp);
        return false;
    }
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        if let Ok(meta) = std::fs::metadata(path) {
            let mut perm = meta.permissions();
            perm.set_mode(perm.mode() | 0o110);
            let _ = std::fs::set_permissions(path, perm);
        }
    }
    // a downloaded executable is quarantined on macOS; clear it so it can run
    if cfg!(target_os = "macos") {
        let _ = Command::new("xattr").args(["-d", "com.apple.quarantine"]).arg(path).output();
    }
    true
}

fn spawn(path: &PathBuf) -> bool {
    let port = managed_port().to_string();
    let direct = Command::new(path)
        .args(["--server", "--nobrowser", "--host", MANAGED_HOST, "--port", &port, "-ngl", "0", "--log-disable"])
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .spawn();
    let child = match direct {
        Ok(c) => c,
        // some hosts must run the APE via a shell
        Err(_) => match Command::new("sh")
            .arg(path)
            .args(["--server", "--nobrowser", "--host", MANAGED_HOST, "--port", &port, "--log-disable"])
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
        {
            Ok(c) => c,
            Err(_) => return false,
        },
    };
    state().lock().unwrap().proc = Some(child);
    true
}

fn wait_started() -> bool {
    let base = format!("http://{}:{}", MANAGED_HOST, managed_port());
    let deadline = Instant::now() + START_TIMEOUT;
    while Instant::now() < deadline {
        {
            let mut st = state().lock().unwrap();
            if let Some(p) = st.proc.as_mut() {
                if let Ok(Some(_)) = p.try_wait() {
                    return false;
                }
            }
        }
        if get_ok(&format!("{}/health", base), Duration::from_secs(2)) || get_ok(&format!("{}/v1/models", base), Duration::from_secs(2)) {
            return true;
        }
        std::thread::sleep(Duration::from_secs(2));
    }
    false
}

/// Stop the model server this app started.
pub fn shutdown() {
    let proc = state().lock().unwrap().proc.take();
    if let Some(mut p) = proc {
        if let Ok(None) = p.try_wait() {
            let _ = p.kill();
            let _ = p.wait();
        }
    }
}

/// One completion from the local model over the
/// OpenAI-compatible API both Ollama and llamafile speak, or "".
pub fn chat(prompt: &str, max_tokens: i64) -> String {
    if let Some(r) = hooks::CHAT.with(|h| h.borrow().as_ref().map(|f| f(prompt, max_tokens))) {
        return r;
    }
    let base = endpoint();
    if base.is_empty() {
        return String::new();
    }
    let model = { let m = state().lock().unwrap().model.clone(); if m.is_empty() { "local".to_string() } else { m } };
    let body = json!({
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.1,
        "max_tokens": max_tokens,
        // a model that ends its turn with a marker of its own: cut there, so no
        // marker reaches a title, which has no sentence for the trim to find
        "stop": ["<end_of_turn>", "<|eot_id|>", "</s>"],
        "stream": false,
    });
    let text = serde_json::to_string(&body).unwrap_or_default();
    let url = format!("{}/v1/chat/completions", base);
    let got = match hooks::POST.with(|h| h.borrow().as_ref().map(|f| f(&url, &text))) {
        Some(r) => r,
        None => bagholder_net::client::request("POST", &url, &[("Content-Type", "application/json")], Some(text.as_bytes()), chat_timeout())
            .map(|r| r.text())
            .map_err(|e| e.to_string()),
    };
    let body = match got {
        Ok(b) => b,
        Err(_) => return String::new(),
    };
    let v: Value = match serde_json::from_str(&body) { Ok(v) => v, Err(_) => return String::new() };
    let content = v
        .get("choices")
        .and_then(|c| c.as_array())
        .and_then(|a| a.first())
        .and_then(|c| c.get("message"))
        .and_then(|m| m.get("content"));
    match content {
        Some(Value::String(s)) => bagholder_model::textrules::trim_space(s).to_string(),
        Some(Value::Null) | None => String::new(),
        Some(other) => bagholder_model::textrules::trim_space(&bagholder_model::value::s(Some(other))).to_string(),
    }
}
